ReactWindows/ReactNative.Core/Bridge/GenericDelegateFactory.cs

```diff
@@ -0,0 +1,64 @@
+using System;
+using System.Reflection;
+
+namespace ReactNative.Bridge
+{
+    public interface IGenericDelegate
+    {
+        object Invoke(object[] args);
+
+        object DelegateTarget
+        {
+            get;
+        }
+    }
+
+    public static class GenericDelegateFactory
+    {
+        static IGenericDelegate CreateNewTypeWrapper(Type[] typeList, object instance, MethodInfo method, bool isFunc)
+        {
+            var numArgs = typeList.Length;
+            if (numArgs == 0)
+            {
+                return new SignatureAction0(instance, method);
+            }
+
+            Type actionT;
+            if (isFunc)
+            {
+                actionT = TypesTemplateFunc.TemplateType[numArgs - 1].MakeGenericType(typeList);
+            }
+            else
+            {
+                actionT = TypesTemplateAction.TemplateType[numArgs].MakeGenericType(typeList);

```

Comment:
This is apt to fail if `numArgs` exceeds the number of delegate types you've generated. In .NET - System.Func and System.Action are capped at 16 generic parameters, but there is no such cap for parameters to a method in .NET. You'll need to add a fallback for 16+ parameters Funcs and 17+ parameter Actions.

Reply:
Here's my suggested approach:
* Leave the `ReflectionReactDelegateFactory` as is
* Create a new class, call it `ReactDelegateFactory`
* Add the implementation involving `IGenericDelegate` there (you may not need an interface, just a base class, maybe `ReactDelegateBase`, since none of these details will be public)
* Add a combined implementation of `IReactDelegateFactory` that uses `ReactDelegateFactory` for methods with 16 or less params (common) and falls back to the `ReflectionReactDelegateFactory` for methods with more than 16 params (uncommon)
* Add unit tests, including methods with more than 16 params.

---
In reply to: [161113590](https://github.com/Microsoft/react-native-windows/pull/1574#discussion_r161113590) [](ancestors = 161113590)

Reply:
on it. thanks for the review!